File: .claude/skills/evaluate-tool/templates/run_tests.m
%% run_tests.m -- Octave test runner for MATPOWER evaluation
%%
%% Copy this file into evaluations/matpower/tests/ to enable test discovery.
%%
%% Usage:
%%   octave tests/run_tests.m                          % run all tests
%%   octave tests/run_tests.m expressiveness           % run one dimension
%%   octave tests/run_tests.m expressiveness test_a1   % run one test

function run_tests(dimension, test_name)
    %% Setup paths
    test_dir = fileparts(mfilename('fullpath'));
    repo_root = fullfile(test_dir, '..', '..', '..', '..');
    data_dir = fullfile(repo_root, 'data', 'networks');

    %% Add MATPOWER to path (assumes setup.sh has been run)
    matpower_dir = fullfile(test_dir, '..', 'matpower8.1');
    if exist(matpower_dir, 'dir')
        addpath(genpath(matpower_dir));
    end

    %% Network file paths
    networks = struct();
    networks.TINY = fullfile(data_dir, 'case39.m');
    networks.SMALL = fullfile(data_dir, 'case_ACTIVSg2000.m');
    networks.MEDIUM = fullfile(data_dir, 'case_ACTIVSg10k.m');
    timeseries.TINY = fullfile(fileparts(data_dir), 'timeseries', 'case39');

    %% Discover test files
    if nargin < 1 || isempty(dimension)
        dimension = '';
    end
    if nargin < 2 || isempty(test_name)
        test_name = '';
    end

    test_files = discover_tests(test_dir, dimension, test_name);

    if isempty(test_files)
        warning('No test files found (dimension=%s, test=%s)', ...
                dimension, test_name);
        return
    end

    fprintf('Found %d test file(s)\n', length(test_files));
    fprintf('==========================================\n');

    %% Run tests
    passed = 0;
    failed = 0;
    errs = 0;

    for i = 1:length(test_files)
        tf = test_files{i};
        [~, fname, ~] = fileparts(tf);
        fprintf('\nRunning: %s\n', fname);
        fprintf('------------------------------------------\n');

        try
            %% Change to test directory so relative paths work
            old_dir = pwd();
            cd(fileparts(tf));

            %% Run the test function
            result = feval(fname);

            cd(old_dir);

            if strcmp(result.status, 'pass') || ...
                    strcmp(result.status, 'qualified_pass')
                fprintf('  PASS (%0.2fs)\n', result.wall_clock_seconds);
                passed = passed + 1;
            else
                fprintf('  FAIL (%0.2fs)\n', result.wall_clock_seconds);
                if isfield(result, 'errors')
                    for j = 1:length(result.errors)
                        fprintf('    Error: %s\n', result.errors{j});
                    end
                end
                failed = failed + 1;
            end
        catch e
            cd(old_dir);
            fprintf('  ERROR: %s\n', e.message);
            errs = errs + 1;
        end
    end

    %% Summary
    fprintf('\n==========================================\n');
    fprintf('Results: %d passed, %d failed, %d errors\n', ...
            passed, failed, errs);
    fprintf('==========================================\n');
end

function test_files = discover_tests(test_dir, dimension, test_name)
    test_files = {};

    if isempty(dimension)
        %% Search all subdirectories
        entries = dir(test_dir);
        for i = 1:length(entries)
            if entries(i).isdir && ~startsWith(entries(i).name, '.')
                dim_path = fullfile(test_dir, entries(i).name);
                files = dir(fullfile(dim_path, 'test_*.m'));
                for j = 1:length(files)
                    if isempty(test_name) || ...
                            contains(files(j).name, test_name)
                        idx = length(test_files) + 1;
                        test_files{idx} = fullfile(dim_path, ...
                                                   files(j).name);
                    end
                end
            end
        end
    else
        dim_path = fullfile(test_dir, dimension);
        if exist(dim_path, 'dir')
            files = dir(fullfile(dim_path, 'test_*.m'));
            for j = 1:length(files)
                if isempty(test_name) || ...
                        contains(files(j).name, test_name)
                    idx = length(test_files) + 1;
                    test_files{idx} = fullfile(dim_path, files(j).name);
                end
            end
        else
            warning('Dimension directory not found: %s', dim_path);
        end
    end

    test_files = sort(test_files);
end

File: .claude/skills/evaluate-tool/templates/runtests.jl
#=
Julia test runner for power-system tool evaluation.

Copy this file into evaluations/<tool>/tests/ to enable test discovery
and execution across dimensions.

Usage:
    julia --project=. tests/runtests.jl                    # run all tests
    julia --project=. tests/runtests.jl expressiveness     # run one dimension
    julia --project=. tests/runtests.jl expressiveness a1  # run one test
=#

using Test
using JSON

# Repository root (3 levels up from evaluations/<tool>/tests/)
const REPO_ROOT = abspath(joinpath(@__DIR__, "..", "..", "..", ".."))
const DATA_DIR = joinpath(REPO_ROOT, "data", "networks")

# Network file paths
const NETWORKS = Dict(
    "TINY" => joinpath(DATA_DIR, "case39.m"),
    "SMALL" => joinpath(DATA_DIR, "case_ACTIVSg2000.m"),
    "MEDIUM" => joinpath(DATA_DIR, "case_ACTIVSg10k.m"),
)
const TIMESERIES = Dict("TINY" => joinpath(dirname(DATA_DIR), "timeseries", "case39"))

# Reference counts for gate validation
const REFERENCE_COUNTS = Dict("TINY" => Dict("buses" => 39, "branches" => 46, "generators" => 10))

"""
    discover_tests(dimension::String="") -> Vector{String}

Find all test files matching the pattern test_*.jl in dimension subdirectories.
If dimension is specified, only search that subdirectory.
"""
function discover_tests(dimension::String="", test_filter::String="")
    test_dir = @__DIR__
    test_files = String[]

    if isempty(dimension)
        # Search all dimension subdirectories
        for entry in readdir(test_dir)
            dim_path = joinpath(test_dir, entry)
            isdir(dim_path) || continue
            for f in readdir(dim_path)
                if startswith(f, "test_") && endswith(f, ".jl")
                    if isempty(test_filter) || contains(lowercase(f), lowercase(test_filter))
                        push!(test_files, joinpath(dim_path, f))
                    end
                end
            end
        end
    else
        dim_path = joinpath(test_dir, dimension)
        if isdir(dim_path)
            for f in readdir(dim_path)
                if startswith(f, "test_") && endswith(f, ".jl")
                    if isempty(test_filter) || contains(lowercase(f), lowercase(test_filter))
                        push!(test_files, joinpath(dim_path, f))
                    end
                end
            end
        else
            @warn "Dimension directory not found: $dim_path"
        end
    end

    sort!(test_files)
    return test_files
end

"""
    run_test_file(path::String) -> Dict

Include and run a test file, capturing its run() output.
"""
function run_test_file(path::String)
    @info "Running: $path"
    try
        include(path)
        # The included file should define a run() function
        if isdefined(Main, :run)
            return Main.run()
        else
            return Dict("status" => "fail", "errors" => ["No run() function defined"])
        end
    catch e
        return Dict("status" => "fail", "errors" => [string(typeof(e), ": ", sprint(showerror, e))])
    end
end

# Parse command-line arguments
dimension_filter = length(ARGS) >= 1 ? ARGS[1] : ""
test_filter = length(ARGS) >= 2 ? ARGS[2] : ""

test_files = discover_tests(dimension_filter, test_filter)

if isempty(test_files)
    @warn "No test files found" dimension=dimension_filter test=test_filter
else
    @info "Found $(length(test_files)) test file(s)"

    @testset "Evaluation Tests" begin
        for tf in test_files
            test_name = basename(tf)
            @testset "$test_name" begin
                result = run_test_file(tf)
                @test result["status"] in ["pass", "qualified_pass"]
                if haskey(result, "errors") && !isempty(result["errors"])
                    @warn "Errors in $test_name" errors=result["errors"]
                end
            end
        end
    end
end
